### PR TITLE
Use Regex for user search

### PR DESCRIPTION
**Pull request by Eyap53**
Now use regex for autocompletion.

Ex : User type "8"

-> results : 8, 8-179, 4-8-12.
But not 18, 80, 12-88, etc...

### CODE REVIEW

**Comment by AlexandrePalo** [comment]
#2

**Comment by Eyap53** [comment]
UPDATE : Doesn't return when family number isn't first

Ex : User type "8"

-> results : 8, 8-179,
But not 18, 80, 12-8, etc...

**Comment by AlexandrePalo** [comment]
Final behavior:

Users: 20-23, 20-1, 2-100, 1-20-29

2 => 2-100
20 => 20-23, 20-1
1 => 1-20-29
20-23 => 20-23
20-2 => empty

**Comment by Eyap53** [comment]
The last commit added the functionnality to Operator sales.

TO DO :
- Use it in other field to search users
- Remove js / comments / code wich is now unused

**Comment by AlexandrePalo** [comment]
5a459cb fix #1 ?

**Comment by AlexandrePalo** [comment]
## Please work with c269884:

When I select the user after autocompletion, "ENTER" submits the form in operator sale module and it shouldn't.
It should only select the user and update the balance in the form.

It can be made with "prevent default" for key handling "ENTER". And the update of the form can be done with "blur()" of the user field for instance  :)

**Comment by Eyap53** [comment]
> What's the aim of 5a459cb ?
Fix the futur balance when you add products (the js function total() is called twice, and were modified only once).

> When I select the user after autocompletion, "ENTER" submits the form in operator sale module and it shouldn't.
It should only select the user and update the balance in the form.

True, i'll work on it.

**Comment by Eyap53** [comment]
Update TODO : 

- [X] Use it in other search-users field
- [X] Remove js / comments / code wich is now unused
- [ ] Add Permissions to Ajax call

**Comment by Eyap53** [comment]
- [X] Use it in other search-users field
- [x] Remove js / comments / code wich is now unused
- [x] Add Permissions to Ajax call

Waiting for review

File: users/views.py [hunk 1]
```diff
@@ -444,18 +444,39 @@ def username_from_username_part(request):
...
 
-    data = 0
-    try:
-        username = request.GET.get('username')
+            data = str(User.objects.get(username=username).balance)
 
-        data = str(User.objects.get(username=username).balance)
 
+        except KeyError:
+            pass
```
Comment (AlexandrePalo) [inline users/views.py]: I think we need to raise something here.
Error is different from just regular 0 balance.

File: users/views.py [hunk 2]
```diff
@@ -444,18 +444,39 @@ def username_from_username_part(request):
...
+        except ObjectDoesNotExist:
+            pass
```
Comment (AlexandrePalo) [inline users/views.py]: Same here.

File: users/views.py [hunk 3]
```diff
@@ -444,18 +444,39 @@ def username_from_username_part(request):
...
+    # except KeyError:
+    #         raise Http404
+    # except ObjectDoesNotExist:
+    #     raise Http404
+
+    # If deactivate
+    # if module.state is False:
+    #     raise Http404
+
+    if operator.has_perm('modules.use_operatorsalemodule'):
```
Comment (AlexandrePalo) [inline users/views.py]: Does this function could be used elsewhere ?
If so, we should change the permission to something like "get_users_balance", don't know hehe

**Comment by AlexandrePalo** [comment]
33e48ee doesn't well work for me on Google Chrome.
It'll work on that :)

File: users/views.py (hunk 3 above)
Comment (Eyap53) [reply users/views.py]: Yeah, i agree 👍

**Comment by AlexandrePalo** [comment]
@Eyap53 review please :)

Todo:
[x] Prevent submit with enter on id_client field.
[x] Update initial and total when select in autocomplete.
[x] Set default initial, total and field when autocomplete option selected is not correct (username not recognize or any error given by balance_from_username). 
[] Auto update initial and total when the page is reloaded. For instance when the form is within an error state, the page is reloaded with all correct field fulfilled: initial and total must be updated too.

**Comment by Eyap53** [comment]
test

**Pull request merged**